File: py/linux_wallpaperengine.py
import configparser
import logging
import os
import subprocess


def kill_wallpaperengine():
    try:
        _ = subprocess.run(["pkill", "-f", "linux-wallpaperengine"], check=True)
        logging.info("已终止linux-wallpaperengine进程")
    except subprocess.CalledProcessError:
        logging.warning("未找到运行中的linux-wallpaperengine进程")


def load_wallpaper(id: str, config: configparser.ConfigParser):
    # wallpaper_path: str =
    os.path.join(config["wallpaper_dir"], id)
    if os.path.isdir(wallpaper_path):
        kill_wallpaperengine()
        time.sleep(1)
        logging.info(f"加载壁纸: {id}")

        base_command = ["linux-wallpaperengine"]
        screen_roots = config.get("screen_root", "").split(",")

        for i, screen_root in enumerate(screen_roots):
            command = base_command.copy()
            command += ["--screen-root", screen_root]

            if i == 0:
                if config.getboolean("silent", fallback=False):
                    command.append("--silent")
                if config.get("volume"):
                    command += ["--volume", config["volume"]]
                if config.getboolean("noautomute", fallback=False):
                    command.append("--noautomute")
                if config.getboolean("no_audio_processing", fallback=False):
                    command.append("--no-audio-processing")
            else:
                command.append("--silent")

            if config.get("scaling"):
                command += ["--scaling", config["scaling"]]
            if config.getboolean("window", fallback=False):
                command.append("--window")
            if config.get("fps"):
                command += ["--fps", config["fps"]]
            if config.get("assets_dir"):
                command += ["--assets-dir", config["assets_dir"]]
            if config.getboolean("screenshot", fallback=False):
                command.append("--screenshot")
            if config.getboolean("list_propertites", fallback=False):
                command.append("--list-propertites")
            if config.get("set_property"):
                set_properties = config.get("set_property").split(",")
                for sp in set_properties:
                    command += ["--set-property", sp]
            if config.getboolean("no_fullscreen_pause", fallback=False):
                command.append("--no-fullscreen-pause")
            if config.getboolean("disable_mouse", fallback=False):
                command.append("--disable-mouse")
            if config.get("clamping"):
                command += ["--clamping", config["clamping"]]

            command.append(id)

            logging.info(f"执行命令: {' '.join(command)}")
            log_output = f"{config['wallpaperengine_log_file']}_in_{screen_root}"
            with open(log_output, "a") as log_file:
                subprocess.Popen(command, stdout=log_file, stderr=log_file)
            time.sleep(1)

        time.sleep(int(config["interval"]))
    else:
        logging.error(f"未找到文件夹: {id}")

File: src/player/linux_wallpaperengine.rs
use std::fs::{self, File};
use std::os::unix::process::CommandExt;
use std::path::Path;
use std::process::{Child, Command, Stdio};
use std::thread::sleep;
use std::time::Duration;

use log::{error, info};
use nix::libc::{kill, killpg, setpgid, SIGKILL, SIGTERM};

use super::config::app_config::PlayCommandConfig;

pub fn load_wallpaper(
    wallpaper_path: &Path,
    wallpaper_id: &str,
    config: &PlayCommandConfig,
) -> Vec<Child> {
    info!("start load_wallpaper");
    if fs::metadata(wallpaper_path).is_ok() {
        info!("loading wallpaper: {}", wallpaper_path.to_str().unwrap());

        let mut pids = vec![];
        for screen_root in &config.screen_root {
            let screen_root: &String = &screen_root;
            let mut command = build_command(config, screen_root, wallpaper_id);

            let log_file = match &(config.log_file) {
                Some(_) => config.log_file.as_ref().unwrap(),
                None => "./linux_wallpaperengine.log",
            };

            info!("Execute command: {:?}", command);
            let process = unsafe {
                command
                    .stdout(Stdio::from(
                        File::create(&log_file).expect("Can not cerea log file!"),
                    ))
                    .stderr(Stdio::from(
                        File::create(&log_file).expect("Can not cerea log file!"),
                    ))
                    .pre_exec(|| {
                        setpgid(0, 0);
                        Ok(())
                    })
                    .spawn()
                    .expect("Error to run command!")
            };
            pids.push(process);
            sleep(Duration::from_millis(500));
        }
        pids
    } else {
        error!("Wallpaper {} not found", wallpaper_path.to_str().unwrap());
        vec![]
    }
}

fn build_command(config: &PlayCommandConfig, screen_root: &String, wallpaper_id: &str) -> Command {
    let mut command = Command::new("linux-wallpaperengine");
    if let Some(scaling) = &config.scaling {
        command.arg("--scaling").arg(scaling);
    }
    if let Some(clamping) = &config.clamping {
        command.arg("--clamping").arg(clamping);
    }
    command.arg("--screen-root").arg(screen_root);

    // [todo]: add silent
    // if i == 0 {
    //     if config.silent.unwrap_or_else(|| false) {
    //         command.arg("--silent");
    //     }
    //     if let Some(volume) = config.volume {
    //         command.arg("--volume").arg(volume.to_string());
    //     }
    //     if config.noautomute.unwrap_or_else(|| false) {
    //         command.arg("--noautomute").arg("--no-audio-processing");
    //     }
    // } else {
    //     command.arg("--silent");
    // }
    command.arg("--silent");

    if config.window.unwrap_or_else(|| false) {
        command.arg("--window");
    }
    if let Some(fps) = config.fps {
        command.arg("--fps").arg(fps.to_string());
    }
    if let Some(assets_dir) = &config.assets_dir {
        command.arg("--assets-dir").arg(assets_dir);
    }
    if config.screenshot.unwrap_or_else(|| false) {
        command.arg("--screenshot");
    }
    if config.list_propertites.unwrap_or_else(|| false) {
        command.arg("--list-propertites");
    }
    // Can not set propertites for all wallpaper
    // if let Some(set_properties) = &config.set_property {
    //     for s in set_properties {
    //         command
    //             .arg("--set-property")
    //             .arg(format!("{}={}", key, value));
    //     }
    // }
    if config.no_fullscreen_pause.unwrap_or_else(|| false) {
        command.arg("--no-fullscreen-pause");
    }
    if config.disable_mouse.unwrap_or_else(|| false) {
        command.arg("--disable-mouse");
    }

    command.arg(wallpaper_id);
    command
}
